### ISSUE TITLE
prefetch to local file - not ~/ncbi/public/sra

### ISSUE THREAD

**Issue by map2085** [ncbi/sra-tools#71]
Is it possible to prefetch an SRA file to a custom directory, rather than the default  (~/ncbi/public/sra)  _without_ changing the SRA toolkit default directory configuration?

e.g.   `prefetch  --outdir .    SRA#  `

Downloading several SRA files simultaneously to my cluster filesystem degrades performance due to high number of writes by prefetch.  I would like to download the SRA file to the local compute node instead, and transfer the SRA file to my filesystem once download is completed.

Changing the default repository is not a solution here, since several different compute nodes would be downloading several different SRA file simultaneously, each to their own local directory (e.g. SGE job specific)  so they all would need to have a different "default repository" directory.

Thank you.

**Comment by kwrodarmer** [comment]
The main technique for doing this today is somewhat kludgy. First, let me say why.

The constraint is that prefetch is a user-level tool working on a personal user's cache. Bombarding a single user's cache from a cluster is generally not a good thing. We have a concept of a "site" repository that allows a single cache to be used from several users/places, but the repository admin still has to take care to manage potential collisions when populating the cache.

In your case, you seem to want to prefetch rather than using https to get data on demand - I presume you're using Aspera? If not, just use https.

For the multiple caches per single user, the way to do this today would be by creating a custom configuration file per job on your cluster. The first part would be to start with a copy of our default configuration file: see https://github.com/ncbi/ncbi-vdb/blob/master/libs/kfg/default.kfg . You'll want to edit the line 38 (as of 4/24/2017) that reads

`/repository/user/main/public/root` = "$(HOME)/ncbi/public"

and replace the path on the right hand side of the assignment operator. There are 2 very important rules to follow or it will not work:
1. the path MUST be in double quotes
2. do not use variable substitution - it doesn't work the way you might think

Replace it with something like

`/repository/user/main/public/root` = "/net/myvol/blah/public"

Assuming that you don't have anything else within your user settings to alter the public user cache and don't have a site-wide configuration, this file should be enough to provide for general toolkit configuration.

The means of getting your toolkit to recognize this file rather than the one it currently sees is by setting and exporting the shell variable `VDB_CONFIG`.

So in summary:

`cp /path/to/default.kfg > /my-cluster-workspace/custom.kfg`  
update the custom.kfg using whatever editor you like - sed, awk, etc.  
`export VDB_CONFIG=/my-cluster-workspace/custom-kfg`  
at this point you should be good to go, and can test the path with `vdb-config`

`vdb-config /repository/user/main`

**Comment by map2085** [comment]
Thank you for the quick reply!
I'm glad to hear there is a workaround.
I will give this a try right now and let you know...

**Comment by kwrodarmer** [comment]
In case it wasn't clear from my previous post, it should be made obvious that you're creating a separate configuration file for each cluster job. So you'll be running `prefetch` from a script that establishes the context.

**Comment by map2085** [comment]
Yes, thank you for clarifying.
I understood your point in the previous post, but it is always best to be explicit anyway.

**Comment by map2085** [comment]
I just deployed this strategy large scale.  
It works very well, thank you for the great advice!

Prefetching the SRA file to the local compute node (i.e. /my/cluster/node) with Aspera and then, once completed, rsync'ing the .sra file to the default path on the filesystem (i.e. ~/ncbi/public) has dramatically lowered the CPU and i/o burden on the filesystem.

In fact, the whole process goes much faster now.

**Comment by kwrodarmer** [comment]
Great to hear.

**Comment by map2085** [comment]
May I add that this technique also works very well for  performing `fastq-dump` for multiple different SRA's simultaneously.

Namely, fastq-dump reads the reference fasta (conventionally stored in ~/ncbi/public/refseq) to convert an .sra file to .fastq.  If several SRA files pertain to the same organism and are encoded using the same RefSeq fasta files, then performing simultaneous/parallel fastq-dumps will burden the filesystem, as multiple processes may be trying to read the same fasta file at the same time.

Using the above strategy, _and also_  copying all of files in ~/ncbi/public/refseq to /my/node/cluster/ncbi/public/refseq, will thereby relieve this filesystem burden as well.

**Comment by kwrodarmer** [comment]
Thanks for that. Yes, it would relieve the burden on the single file system. Another thing that we recommend for external references is to set up a site repository for them and to simply copy all of the sequences in bulk before starting your jobs. These sequences can be found at https://ftp-trace.ncbi.nlm.nih.gov/sra/refseq/ . With this approach, you won't have any downloads to perform during prefetch or processing.
